Fixes to the sort timestamp of messages

iequidoo opened the pull request:
See commit messages.

There's no test yet, verified by hand only. Need to come up with more simple scenario than in #5088.

Fixes #5088

Hocuri commented on src/chat.rs:
```diff
@@ -1291,6 +1295,75 @@ impl ChatId {
+            context
+                .sql
+                .query_get_value(
+                    "SELECT MAX(timestamp) FROM msgs \
+                    WHERE chat_id=? AND hidden=0 AND state>? AND state!=?",
+                    (self, MessageState::InFresh, MessageState::OutDraft),
```
Why check whether it's a draft after we already checked whether it's hidden? Drafts are set as `hidden`, otherwise they would be shown in the chat, just as location messages, reactions, etc.

Hocuri commented on src/chat.rs:
```diff
@@ -1291,6 +1295,75 @@ impl ChatId {
+            context
+                .sql
+                .query_get_value(
+                    "SELECT MAX(timestamp) FROM msgs WHERE chat_id=? AND state!=?",
+                    (self, MessageState::OutDraft),
```
Intuitively, here we should also check whether it's a hidden message, not only whether it's a draft

> But let hidden messages sort together with protection messages because hidden messages also can be or not be verified.

But hidden messages aren't shown to the user, and the sort timestamp is about in which order messages are shown to the user?

iequidoo replied:
When it comes to protection messages, the sort timestamp is not only about the order of messages shown to the user, but also about which messages are verified. Hidden messages such as location ones also can be verified. If we allow them to have a relaxed ordering with protection messages, we lose information about their verification status. This information can be useful to display some additional info in the chat, maybe not now but in the future DC versions, so i'd like to preserve it

iequidoo replied:
Or i'm wrong and `Param::GuaranteeE2ee` is for that?

iequidoo replied:
But as i see from the code, a message can be encrypted (and displayed with a padlock), but not verified, even in verified chats. Also checked this in my DC Desktop

Hocuri replied:
> Or i'm wrong and `Param::GuaranteeE2ee` is for that?

No it's not

> But as i see from the code, a message can be encrypted (and displayed with a padlock), but not verified (be in an unverified section), even in verified chats. Also checked this in my DC Desktop

Yes

----

> This information can be useful to display some additional info in the chat, maybe not now but in the future DC versions, so i'd like to preserve it

Usually when one tries to plan for a possible future like this, it never happens.

That said, it also isn't bad as it is right now - it violates the principle of least astonishment a little bit because here we only check for whether it's a draft and a bit further down we also check for whether it's hidden, and the reason for that isn't clear for the reader, but that's all.

iequidoo replied:
Added a comment on this so that it is easier to understand for readers

Hocuri commented on src/receive_imf.rs:
```diff
@@ -826,9 +815,9 @@ async fn add_parts(
-                        let sort_timestamp =
-                            calc_sort_timestamp(context, sent_timestamp, chat_id, true, incoming)
-                                .await?;
+                        let sort_timestamp = chat_id
+                            .calc_protection_msg_sort_timestamp(context, mime_parser.timestamp_sent)
+                            .await?;
                         chat_id
                             .set_protection(context, new_protection, sort_timestamp, Some(from_id))
                             .await?;
```
Nit, feel free to ignore:

We're almost always calling `calc_protection_msg_sort_timestamp()` right before `set_protection()` - except for `create_group_chat()`, where it wouldn't hurt (and one database request whenever a group is created isn't a lot - what we should avoid is more database requests every time a message is received).

So, `calc_protection_msg_sort_timestamp()` _could_ be moved into `set_protection()`. And since it would then be used only a single time, it could be inlined. Both these steps would reduce code complexity I think.

iequidoo replied:
Added `ChatId::set_protection_for_timestamp_sent()` for this. Better ideas for the function name are welcome

iequidoo replied:
Another option is to add an enum for timestamps with variants `Sent(i64), Sort(i64)`, but not sure it's better

Hocuri replied:
For the name, you could rename `set_protection()` to `set_protection_for_timestamp_sort()` since it's an internal function now that's only used within `chat.rs` and doesn't need a nice name. And rename `set_protection_for_timestamp_sent()` to `set_protection()` since it's the function that everyone should use when in doubt.

In any case, you should inline `calc_protection_msg_sort_timestamp()` since it's a two-line function that is used only a single time.

iequidoo replied:
Done

iequidoo commented:
Rebased and added [fix: Add forgotten checks for Config::VerifiedOneOnOneChats](https://github.com/deltachat/deltachat-core-rust/pull/5094/commits/f1c8cd10bb14b36413d9393f3c578856c9eac19c)

Hocuri commented on src/securejoin.rs:
```diff
@@ -616,12 +616,17 @@ pub(crate) async fn observe_securejoin_on_other_device(
+                        mime_message.timestamp_sent,
+                    )
+                    .await?;
+                }
```
The protected status of chats is still maintained even if `Config::VerifiedOneOnOneChats` is not set, so it was correct not to check for `Config::VerifiedOneOnOneChats`. `Config::VerifiedOneOnOneChats` only means that the UI has support for it (in hindsight, this wasn't a good name for this config). Also see

https://github.com/deltachat/deltachat-core-rust/blob/582541634f806c3044d41cc54e6b0d0e6f8cb081/src/receive_imf.rs#L801-L824

Also, this doesn't have anything to do with sort timestamps, and in general, it's better to keep the PRs as small as possible in order to make reviews faster. It's a bit unfortunate that I have so limited time right now, which makes reviews take a long time, but you can help by splitting up your PRs as much as possible - it's way easier to reason about changes if I can reason about them one at a time.

iequidoo replied:
Ok, let's not discuss this here. I added this commit because it textually depends on the previous ones. Probably the right way in such cases is to create another one PR from one of my branches to another (to avoid self-conflicts). I will remove it, but let's leave `refactor: Move calc_{protection_msg_,}sort_timestamp() to 'impl ChatId'` here, it's just a minor refactoring

iequidoo replied:
Moved to https://github.com/deltachat/deltachat-core-rust/pull/5107

Hocuri commented:
Any progress on trying to add a test?

iequidoo commented:
> Any progress on trying to add a test?

There's the same problem as in `test_create_protected_grp_multidev()` here. I tried to add a Rust test that checks that the protection message has the correct timestamp, but then i need to add large sleep()s to reproduce the bug with a sufficient probability, 3 secs as per my experiments. Otherwise the test has no sense -- it would always pass.

Another option is to add a Python test, but as we have less control over sent and received messages there, we can't write a test just checking a protection message timestamp and need to write some more high-level test, probably repeating the original @gerryfrancis's scenario, but it would be more complex and cumbersome than other tests we have now.

The solution i can imagine here is to mock the system clock for the Rust tests, then we can write any tests checking timestamps of messages w/o adding unnecessary sleep()s.

Hocuri reviewed (approved):
Nice stuff! Sad about the missing test, maybe we can add a way to mock the time at some point
